### PR TITLE
Added query_config parameter to read_gbq

### PR DESCRIPTION
**Pull request by necnec**
Now more complicated queries could be processed.

### CODE REVIEW

**Comment by jreback** [comment]
cc @parthea 

accepting kwargs is a more general way of doing this. I don't really want to have to keep adding specific keywords.

Futher would need some tests.

**Comment by necnec** [comment]
@jreback thanks for your notes. I've changed parameters to kwargs style.
Could you give me some examples how to make tests and publish it as you mentioned. I've done it on my local but I think it is not enough.

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/blob/master/pandas/io/tests/test_gbq.py

for tests and http://pandas-docs.github.io/pandas-docs-travis/contributing.html#running-google-bigquery-integration-tests

**Comment by necnec** [comment]
@jreback do I must write some additional test in https://github.com/pandas-dev/pandas/blob/master/pandas/io/tests/test_gbq.py? As I see current version is passed by bot successfully.
If yes, do I need just write test and push it, or I must make these manipulations http://pandas-docs.github.io/pandas-docs-travis/contributing.html#running-google-bigquery-integration-tests before?

**Comment by parthea** [comment]
> do I must write some additional test in https://github.com/pandas-dev/pandas/blob/master/pandas/io/tests/test_gbq.py? As I see current version is passed by bot successfully.

Correct. There should be new unit test(s) added in `test_gbq.py`

>If yes, do I need just write test and push it, or I must make these manipulations http://pandas-docs.github.io/pandas-docs-travis/contributing.html#running-google-bigquery-integration-tests before?

Please follow the steps in the link mentioned. Ideally this is done before you push. You only need to do this once and hopefully you won't touch it again, except to change credentials if needed. It should only take 5-10 minutes. Separately, please let me know if any part of the instructions is unclear (I put the instructions together).

**Comment by necnec** [comment]
@parthea thank you for your notes.
I've added 2 new tests based on query config. 
I'm not sure about ```test_query_with_parameters``` because  as I see ```parameters``` is an experimental feature now.

BQ testing instruction is quite good, although I have problems with credential and my build crashes on my own Travis-CI tests. I've got an exception 
```InvalidPrivateKeyFormat: Private key is missing or invalid. It should be service account private key JSON (file path or string contents) with at least two keys: 'client_email' and 'private_key'```
although I've put my JSON in single quotes

So, is it enough for my pull-request?

File: doc/source/whatsnew/v0.19.0.txt [hunk 1]
```diff
@@ -379,6 +379,7 @@ Google BigQuery Enhancements
 
 - The :func:`read_gbq` method has gained the ``dialect`` argument to allow users to specify whether to use BigQuery's legacy SQL or BigQuery's standard SQL. See the :ref:`docs <io.bigquery_reader>` for more details (:issue:`13615`).
 - The :func:`~DataFrame.to_gbq` method now allows the DataFrame column order to differ from the destination table schema (:issue:`11359`).
+- The :func:`read_gbq` method now allows query configuration preferences
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: move to 0.19.2

File: pandas/io/gbq.py [hunk 2]
```diff
@@ -682,6 +686,13 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
 
         .. versionadded:: 0.19.0
 
+    **kwargs: Arbitrary keyword arguments
```
Comment (jreback) [inline pandas/io/gbq.py]: this is just like using a named argument. what I would like to see is something like:

```
config = {'query' : ....}
read_gbq(.....,configuration=config)
```
then just add these keys directly.

Comment (necnec) [inline pandas/io/gbq.py]: @jreback Hmm, in this case I can use ´´´read_gbq´´´  for loading files, isn't it? I mean:
´´´
config = {'load' : ....}
´´´
and function ´´´read_gbq´´´ could use not only for reading queries and parameter ´´´query´´´ seems unnecessary. Could you give me some more motivation about configuration parameter? 
I think bigquery jobs like ´´´load´´´, ´´´copy´´´ are outside of pandas philosophy?

**Comment by jorisvandenbossche** [comment]
@necnec sorry to ask you to move it again, but as this is not a critical fix or enhancement, let's keep this for 0.20, so can you move the whatsnew notice to v0.20.0.txt?

File: doc/source/whatsnew/v0.20.0.txt [hunk 3]
```diff
@@ -83,3 +83,10 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+
+.. _whatsnew_0200.gbq:
+
+Google BigQuery Enhancements
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: doesn't need a sub-section, just add to enhancements

File: doc/source/whatsnew/v0.20.0.txt [hunk 4]
```diff
@@ -83,3 +83,10 @@ Performance Improvements
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+- The :func:`read_gbq` method now allows query configuration preferences
```
Comment (necnec) [inline doc/source/whatsnew/v0.20.0.txt]: @jreback 
Sorry, but I have some newbie questions
if there is no issue for that feature do I need to create one now? And do I must add tiny example to ```whatsnew``` file or to ```issue``` file?

File: pandas/io/gbq.py (hunk 2 above)
Comment (jreback) [inline pandas/io/gbq.py]: not sure what you mean. the point here is that if someone wants to specify a configuration option, then they can just pass it thru as a a dict structure. I don't want to have to change pandas code again when someone want 'another' option. These should just pass thru.

**Comment by jorisvandenbossche** [comment]
I think it would be good to add some documentation about this in the tutorial docs? (http://pandas.pydata.org/pandas-docs/stable/io.html#google-bigquery-experimental)

File: pandas/io/gbq.py (hunk 2 above)
Comment (necnec) [inline pandas/io/gbq.py]: @jreback sorry for big response but it doesn't still clear for me:
1) The things I worry about are arguments ```query``` and ```dialect```. It is become redunt.
```
config = {
  'query' : {
    'query':  'select 1',
    'useLegacySql': dialect == 'legacy'
  }
}
read_gbq(query, dialect, configuration=config)
```
 So, I am going to use such logic: if query not specified in config I will use query parameter of read_gbq. But if query or dialect specified in config do I must throw an exception?

2) As I got your idea configuration parameter should be like

```
config = {
  'query' : {
    "useQueryCache": False,
  }
}
```
Not like:
```
config = {
  '"useQueryCache": False
}
```
And in this case I could pass:
```
config = {'load' : ....}
```
what I should do with query parameter of function ```read_gbq```? just skip it inside the logic.

**Comment by jreback** [comment]
you can just use a passed in config dict to update the query

this should be like 2 lines of code for the change

File: doc/source/whatsnew/v0.20.0.txt [hunk 5]
```diff
@@ -50,6 +50,7 @@ Other enhancements
 - ``Series.sort_index`` accepts parameters ``kind`` and ``na_position`` (:issue:`13589`, :issue:`14444`)
 
 - ``pd.read_excel`` now preserves sheet order when using ``sheetname=None`` (:issue:`9930`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: can you add a mini section to the docs and put a pointer here (or is the doc-string enough)?

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: add this PR number as the issue number

File: pandas/io/gbq.py [hunk 6]
```diff
@@ -395,6 +395,13 @@ def run_query(self, query):
                 }
             }
         }
+        configuration = kwargs.get('configuration')
+        if configuration is not None:
```
Comment (jreback) [inline pandas/io/gbq.py]: might be better to setup ``job_data['configuration']`` as a default_dict?

File: pandas/io/tests/test_gbq.py [hunk 7]
```diff
@@ -707,10 +707,63 @@ def test_invalid_option_for_sql_dialect(self):
...
+                        },
+                        "parameterValue": {
+                            "value": 2
+                        }
+                    }
+                ]
+            }
+        }
+        # Test that an invalid query without query config
+        with tm.assertRaises(ValueError):
```
Comment (jreback) [inline pandas/io/tests/test_gbq.py]: why is this necessary? I thought ``configuration`` is an optional paramter? when is it needed

Comment (necnec) [inline pandas/io/tests/test_gbq.py]: @jreback Yes, configuration is optional. But this unit test is very special. It processes query with parameters. And in this case you must pass parameters values in configuration.

I've made 2 unit tests. So if you think this test if very special I can remove that.

Comment (necnec) [inline pandas/io/tests/test_gbq.py]: @jreback so I don't need to change anything here?

**Comment by jreback** [comment]
lgtm. 


@parthea ok?

File: doc/source/whatsnew/v0.20.0.txt [hunk 8]
```diff
@@ -51,6 +51,8 @@ Other enhancements
...
 
+- ``pd.io.gbq.read_gbq`` method now allows query configuration preferences (:issue:`14742`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.20.0.txt]: pd.io.gbq.read_gbq -> pd.read_gbq (available as top-level function)

File: pandas/io/gbq.py [hunk 9]
```diff
@@ -682,6 +692,15 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
 
         .. versionadded:: 0.19.0
 
+    **kwargs: Arbitrary keyword arguments
```
Comment (jorisvandenbossche) [inline pandas/io/gbq.py]: Space after 'kwargs' (before the colon, this is a numpydoc peculiarity)

Comment (jorisvandenbossche) [inline pandas/io/gbq.py]: Also, the ``**`` in the kwargs will give an error when building the docs (as this is used to make something bold). 

If you want to keep them, you need to make the docstring a raw string by adding a 'r' at the front (so ``r"""``)

File: pandas/io/gbq.py [hunk 10]
```diff
@@ -682,6 +692,15 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
...
+        configuration (dict): query config parameters for job processing.
+            For example:
```
Comment (jorisvandenbossche) [inline pandas/io/gbq.py]: No extra indentation here (compared to the previous line)

File: pandas/io/gbq.py [hunk 11]
```diff
@@ -682,6 +692,15 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
...
+                configuration = {'query': {'useQueryCache': False}}
```
Comment (jorisvandenbossche) [inline pandas/io/gbq.py]: there need to be a blank line before and after this line (again, this are some sphinx (rst) syntax things)

File: pandas/io/gbq.py [hunk 12]
```diff
@@ -682,6 +692,15 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
...
+            For more information see `BigQuery SQL Reference
```
Comment (jorisvandenbossche) [inline pandas/io/gbq.py]: Indent this also one level less (same level as the "For example .."

**Review by jorisvandenbossche — COMMENTED** [review]
Two questions:

- if this is to add a specific keyword `configuration`, why not just documenting this keyword instead of putting it under 'kwargs'? (maybe discussed before)

- Can you add some documentation about this feature in the tutorial docs in io.rst (the gbq section)

File: pandas/io/gbq.py [hunk 13]
```diff
@@ -385,16 +385,25 @@ def run_query(self, query):
...
+                # 'allowLargeResults', 'createDisposition',
+                # 'preserveNulls', destinationTable, useQueryCache
             }
         }
+        configuration = kwargs.get('configuration')
+        if configuration is not None:
+            if 'query' in configuration:
+                job_config['query'].update(configuration['query'])
+            else:
+                job_config = configuration
```
Comment (parthea) [inline pandas/io/gbq.py]: Can you show an example of when `'query'` is not in `configuration`? The unit tests all have `'query'` included in `configuration`.

I think we should ensure that users specify `'query'` in the configuration to prevent users from trying `'load'`, `'extract'` or `'copy'` which are outside of the scope of this PR.

Could we throw an exception or ignore the configuration if `'query'` is not in `configuration`?

File: pandas/io/gbq.py (hunk 10 above)
Comment (parthea) [inline pandas/io/gbq.py]: I think it will be helpful to have an example under the `Google BigQuery (Experimental)` heading in https://github.com/pandas-dev/pandas/blob/master/doc/source/io.rst?

**Review by parthea — COMMENTED** [review]
Overall, this LGTM, but there are a couple of minor notes.

File: pandas/io/gbq.py (hunk 13 above)
Comment (necnec) [inline pandas/io/gbq.py]: @parthea I think there is one problem about throwing exception with ```useLegacySql```. Current logic force ```useLegacySql``` to be specified as ```'legacy'``` or ```'standart'```. 
So do I need throw exception always when I specify ```useLegacySql``` in  ```configuration```? 
I think rewriting this parameter from configuration is better solution (current version).

File: doc/source/whatsnew/v0.20.0.txt (hunk 8 above)
Comment (necnec) [inline doc/source/whatsnew/v0.20.0.txt]: @jorisvandenbossche about ```kwargs``` instead of ```configuration``` you can find there: https://github.com/pandas-dev/pandas/pull/14742#issuecomment-262965915
I think @jreback has better explanation why we use ```kwargs``` not ```configuration```

File: doc/source/io.rst [hunk 14]
```diff
@@ -4562,6 +4562,20 @@ destination DataFrame as well as a preferred column order as follows:
...
+
+.. code-block:: python
+
+   config = {
+      'query': {
+        "useQueryCache": False
+      }
+   }
+   data_frame = pd.read_gbq('SELECT * FROM test_dataset.test_table',
+                             configuration=config, projectid)
```
Comment (jreback) [inline doc/source/io.rst]: call this config (shorter), rather than configuration

File: pandas/io/tests/test_gbq.py [hunk 15]
```diff
@@ -707,10 +707,95 @@ def test_invalid_option_for_sql_dialect(self):
                          private_key=_get_private_key_path())
 
         # Test that a correct option for dialect succeeds
-        # to make sure ValueError was due to invalid dialect
+        # to make sure ValueError was due to invalid  dialect
```
Comment (parthea) [inline pandas/io/tests/test_gbq.py]: Please revert the extra space before dialect here

**Review by parthea — COMMENTED** [review]
All gbq unit tests pass locally. A couple of minor observations.

**Review by parthea — APPROVED** [review]

File: doc/source/io.rst (hunk 14 above)
Comment (jorisvandenbossche) [inline doc/source/io.rst]: Although `config` is shorter, it is 'configuration' that is used by gbq (https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration), so shouldn't we follow gbq's terminology here?

**Comment by jorisvandenbossche** [comment]
@necnec 

> @jorisvandenbossche about kwargs instead of configuration you can find there: #14742 (comment)
I think @jreback has better explanation why we use kwargs not configuration


kwargs can be used to pass through a bunch of keywords, but this is a single specific keyword that pandas implements to hold those kwargs, so I think we should document it like that.


Further, the user can now pass other keys in the `config` object besides 'query', and those will be ignored but no error/warning is raised?

**Comment by necnec** [comment]
@jorisvandenbossche 
> kwargs can be used to pass through a bunch of keywords, but this is a single specific keyword that pandas implements to hold those kwargs, so I think we should document it like that.

Sorry, I haven't got your idea. Do you think we should use ```config``` argument instead of ```kwargs``` ? It was my first version but @jreback reasonably ask me to change it to ```kwargs```. I'm ready to add some more documentation if you show me where.

>Further, the user can now pass other keys in the config object besides 'query', and those will be ignored but no error/warning is raised?

```config``` allows only 'query' options. Otherwise, it throws an exception. Unit test named ```test_configuration_without_query``` check that

**Comment by jorisvandenbossche** [comment]
> config allows only 'query' options. Otherwise, it throws an exception. Unit test named test_configuration_without_query check that

Yes, but if you pass a config object with 'query' and such a 'copy' key, the 'copy' key will be silently ignored?

**Comment by necnec** [comment]
@jorisvandenbossche Yes it will be ignored. Do you think it should throws an exception in this case?

**Comment by jorisvandenbossche** [comment]
@necnec IMO, yes, I would raise an error, that is more informative to the user (or at least raise a warning)

**Comment by necnec** [comment]
@jorisvandenbossche I've added exception if you pass 2 types of job in config. But I haven't added additional unit test.
So after test this feature on my local I receive exception something like this:
```ValueError: Only one job type must be specified, but given query,load```
I'm not sure that message looks well so I'll happy if you help me to make it more clear.

**Comment by jreback** [comment]
@necnec the reason I thought it a nice idea to allow arbitrary kwargs to be passed is that I don't want pandas having to add lots of kwargs to the signature when they are simply passed thru.

does this assertion still hold? is this useful?

**Comment by necnec** [comment]
Yes, I think adding kwargs is a good idea to pass arbitrary keys

File: doc/source/io.rst [hunk 16]
```diff
@@ -4649,6 +4649,20 @@ destination DataFrame as well as a preferred column order as follows:
                              index_col='index_column_name',
                              col_order=['col1', 'col2', 'col3'], projectid)
 
+
```
Comment (jreback) [inline doc/source/io.rst]: add starting in 0.20.0 (or you can add a versionadded tag)

File: doc/source/io.rst [hunk 17]
```diff
@@ -4649,6 +4649,20 @@ destination DataFrame as well as a preferred column order as follows:
...
+You can specify the query config as parameter
+
+.. code-block:: python
```
Comment (jreback) [inline doc/source/io.rst]: say *why* this is useful as well. If you have a doc-link to things that you might want to pass here, pls add it.

File: pandas/io/gbq.py [hunk 18]
```diff
@@ -385,16 +385,33 @@ def run_query(self, query):
...
+
+        job_config = {
+            'query': {
+                'query': query,
+                'useLegacySql': self.dialect == 'legacy'
+                # 'allowLargeResults', 'createDisposition',
+                # 'preserveNulls', destinationTable, useQueryCache
             }
         }
+        config = kwargs.get('config')
```
Comment (jreback) [inline pandas/io/gbq.py]: can you add a comment on what you are doing here (and why)

File: pandas/io/gbq.py [hunk 19]
```diff
@@ -682,6 +700,17 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
 
         .. versionadded:: 0.19.0
 
+    **kwargs : Arbitrary keyword arguments
+        config (dict): query config parameters for job processing.
+        For example:
+
+            config = {'query': {'useQueryCache': False}}
+
+        For more information see `BigQuery SQL Reference
```
Comment (jreback) [inline pandas/io/gbq.py]: yes this is a good reference, add this above where I indicated

**Comment by jreback** [comment]
@necnec just a couple of doc comments. ping when green.

File: pandas/io/gbq.py [hunk 20]
```diff
@@ -682,6 +700,17 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
...
+            <https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.query>`
```
Comment (jorisvandenbossche) [inline pandas/io/gbq.py]: no indentation relative to "For more ...) is needed here (otherwise possibly will give errors when building the docs)

**Comment by jorisvandenbossche** [comment]
@jreback Repeating my question from above: I know you asked to rename the kwarg from configuration to config, but as this is to specifically pass the 'configuration' from GBQ terminology, shouldn't we rather keep it consistent with their naming and use 'configuration' as well?

**Comment by jreback** [comment]
I just though ``config`` is more in-line with what is actually used in the JSON (and a bit shorter). If we are actually using ``configuration`` (to pass the data) then that would be fine too.

**Comment by jorisvandenbossche** [comment]
@jreback yes, the key in the dict / json that is passed is actually `'configuration'`, therefore I would use that.

**Comment by jreback** [comment]
@necnec ok pls make the adjustment from ``config`` -> ``configuration`` as @jorisvandenbossche indicates. ping on green.

**Comment by jreback** [comment]
thanks @necnec 

sometimes these things takes time. thanks for the patience.

**Comment by necnec** [comment]
Thank you too for this experience! I'm glad to be participated in it.
